settings: Use em for header instead of px.

evykassirer opened the pull request:
Sorry that screenshots aren't exactly aligned, I still don't have software set up to do that.

---

12px

| before             |  after |
:-------------------------:|:-------------------------:
<img width="906" alt="image" src="https://github.com/user-attachments/assets/4f3beea4-9d34-4850-ae2e-a087303cad46" /> | <img width="1030" alt="image" src="https://github.com/user-attachments/assets/810fe2f6-d8c3-4068-9921-aa94d3192597" />
<img width="568" alt="image" src="https://github.com/user-attachments/assets/df2de957-db1e-4483-b0f7-afd51ee41518" /> | <img width="569" alt="image" src="https://github.com/user-attachments/assets/7f8d9661-b486-48ea-90ea-c74d043bc4b7" />
<img width="573" alt="image" src="https://github.com/user-attachments/assets/3f5c10b5-23b4-4b35-ac18-c37e1d413314" /> | <img width="565" alt="image" src="https://github.com/user-attachments/assets/ff9e55a9-02b9-40fa-aba3-23b4b13f17ba" />


14px

| before             |  after |
:-------------------------:|:-------------------------:
<img width="903" alt="image" src="https://github.com/user-attachments/assets/787ead62-b2ae-4bc2-a7d7-1a27c4393788" /> | <img width="1029" alt="image" src="https://github.com/user-attachments/assets/cadf8ad3-62a9-4b59-915d-3e106399322c" />
<img width="573" alt="image" src="https://github.com/user-attachments/assets/e8c8ca2b-7597-41af-b889-9afd8edc66b4" /> | <img width="566" alt="image" src="https://github.com/user-attachments/assets/0497b00e-7c91-4dd6-9ae1-87625402d603" />
<img width="575" alt="image" src="https://github.com/user-attachments/assets/7a8558ec-7ce1-4606-a89b-812ed9275ef8" /> | <img width="568" alt="image" src="https://github.com/user-attachments/assets/9f97bdb9-e943-4100-97b9-3fac097e6788" />

16px 

| before             |  after |
:-------------------------:|:-------------------------:
<img width="1030" alt="image" src="https://github.com/user-attachments/assets/fbbfbd5e-0398-40ee-b40a-ec6c4f1d912d" /> | <img width="1036" alt="image" src="https://github.com/user-attachments/assets/11d313fd-4136-4762-974f-8002ea4a7c08" />
<img width="576" alt="image" src="https://github.com/user-attachments/assets/6e23e19c-5227-4814-a80a-1f856a7c2e21" /> | <img width="567" alt="image" src="https://github.com/user-attachments/assets/98204c0d-1ca9-4d1d-9d34-e84b809caedb" />
<img width="573" alt="image" src="https://github.com/user-attachments/assets/fd45bf6a-06dd-4b03-942e-b6669db135d1" /> | <img width="566" alt="image" src="https://github.com/user-attachments/assets/a66c911d-fc3f-462d-ac30-1565426dca38" />


18px 

| before             |  after |
:-------------------------:|:-------------------------:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/bc5c14c3-8e3f-4641-969a-9ca805325ffa" /> | <img width="1033" alt="image" src="https://github.com/user-attachments/assets/76418be6-7ac8-483d-8297-8a2e62e23f76" />
<img width="575" alt="image" src="https://github.com/user-attachments/assets/d2f06efc-742d-4ff8-8fe2-2b06d64d8fa3" /> | <img width="567" alt="image" src="https://github.com/user-attachments/assets/d711381d-dabd-45cc-9817-df63ccd91c27" />
<img width="572" alt="image" src="https://github.com/user-attachments/assets/1a24e10b-8caa-40d8-859e-7c6f433dd127" /> | <img width="569" alt="image" src="https://github.com/user-attachments/assets/3430e05b-34af-4096-99c9-a100e812dcf7" />


20px 

| before             |  after |
:-------------------------:|:-------------------------:
<img width="901" alt="image" src="https://github.com/user-attachments/assets/f75496af-8160-454d-8428-30643eee9f01" /> | <img width="1030" alt="image" src="https://github.com/user-attachments/assets/6c82ea6b-7d78-4f79-998b-4d56f66b03e9" />
<img width="575" alt="image" src="https://github.com/user-attachments/assets/34002cab-0e00-46cc-a5a9-582b35bee9d4" /> | <img width="566" alt="image" src="https://github.com/user-attachments/assets/6bd28cc5-d68f-40a7-9680-5746f17eb295" />
<img width="570" alt="image" src="https://github.com/user-attachments/assets/918b1435-9a3c-46cb-8c5c-664cda52ebc7" /> | <img width="566" alt="image" src="https://github.com/user-attachments/assets/91cf9fcb-ad7b-41f0-ab42-51c3c4dc9275" />

timabbott commented:
Let's make it actually look right while we're here.

evykassirer commented on web/styles/settings.css:
```diff
@@ -1308,7 +1308,9 @@ label.preferences-radio-choice-label {
-            padding: 6px;
+            display: flex;
+            align-items: center;
+            justify-content: center;
```
The padding isn't necessary anymore because the contents are centered with flex. Previously they were un-centeredly placed with padding.

evykassirer commented on web/styles/settings.css:
```diff
@@ -1447,14 +1452,13 @@ label.preferences-radio-choice-label {
             text-align: center;
             font-size: 1.1em;
             line-height: 1;
-            margin: 15px;
```
The margin isn't necessary anymore because the contents are centered with flex. Previously they were un-centeredly placed with margin.

evykassirer commented on web/styles/settings.css:
```diff
@@ -1447,14 +1452,12 @@ label.preferences-radio-choice-label {
         .exit {
             font-weight: 600;
             position: absolute;
-            top: 10px;
```
This isn't necessary anymore because the contents are vertically centered with flex. The `position:absolute` and `right` are still necessary to place the `x` on the right

evykassirer commented on web/styles/settings.css:
```diff
@@ -1463,9 +1466,8 @@ label.preferences-radio-choice-label {
         .exit-sign {
             float: right;
             position: relative;
-            top: 1px;
```
This isn't necessary anymore because the contents are vertically centered with flex.

karlstolley replied:
Do we still need the float and position declarations then either? If `.exit-sign` is a flex item, then the answer should be No.

evykassirer replied:
Good call! I ended up removing several things:

```diff
         .exit-sign {
-            float: right;
-            position: relative;
-            margin-left: 3px;
             font-size: 1.5em;
-            line-height: 1;
             font-weight: 600;
             cursor: pointer;
         }
```

evykassirer commented on web/styles/settings.css:
```diff
@@ -1786,7 +1788,25 @@ label.preferences-radio-choice-label {
-            display: block;
+            display: flex;
+            justify-content: space-between;
+            height: 45px;
```
This works with all font sizes

karlstolley replied:
Let's note that then, so we don't confuse ourselves in the future as to why this height is expressed in pixels.

evykassirer commented on web/styles/settings.css:
```diff
@@ -1786,7 +1788,25 @@ label.preferences-radio-choice-label {
+                cursor: pointer;
+                transition: none 0.3s ease;
+                transition-property: opacity, transform;
+            }
```
I moved this out of `subscriptions.css` and removed the relative positioning stuff.

evykassirer commented on web/templates/settings_overlay.hbs:
```diff
@@ -5,7 +5,6 @@
         <div class="exit">
             <span class="exit-sign">&times;</span>
         </div>
-        <div class="clear-float"></div>
```
This seemed only relevant before switching to flex.

evykassirer commented:
Screenshots are updated, this is ready for review again!

alya commented:
Can we adjust the height of the header area too? It feels pretty squished vertically at 20px.

evykassirer commented:
Easy fix! Pushing that now.

20px:

<img width="1006" alt="image" src="https://github.com/user-attachments/assets/5781284d-4558-4f2c-9743-4894fe2cbc25" />

16px

<img width="1001" alt="image" src="https://github.com/user-attachments/assets/8c1f9c3b-f4d1-4d38-afdd-0d0dea918361" />

14px

<img width="1005" alt="image" src="https://github.com/user-attachments/assets/02f8bb30-a362-45b9-8aa9-de9f06f0d3ea" />


12px

<img width="1002" alt="image" src="https://github.com/user-attachments/assets/643fcd71-70f8-4772-be7a-8a3c1510c77b" />

alya commented:
Cool, that looks good to me in manual texting. @sahil839 could you please review?

sahil839 commented on web/styles/settings.css:
```diff
@@ -1786,7 +1788,25 @@ label.preferences-radio-choice-label {
+            height: 45px;
+
+            .fa-chevron-left {
+                position: static;
```
Can we remove the relative positionning set for this, since `display` is set to `none` for larger widths and for smaller widths the relative positioning properties are overridden here.

evykassirer replied:
Yes! Good call. Adding this diff to the main commit:

```diff
diff --git a/web/styles/settings.css b/web/styles/settings.css
index b3656f5ec9..53f749c751 100644
--- a/web/styles/settings.css
+++ b/web/styles/settings.css
@@ -1444,9 +1444,6 @@ label.preferences-radio-choice-label {
 
             .fa-chevron-left {
                 float: left;
-                position: relative;
-                top: 14px;
-                left: 10px;
             }
         }
```

evykassirer replied:
Done, thanks!

sahil839 reviewed (commented):
Posted one comment.

Is it easy to increase the left panel width with font size such that the tabs end inside the right border of the left panel at 20px?
![image](https://github.com/user-attachments/assets/dab4f5e7-3fe5-4051-a1c4-68d63fb0b486)

evykassirer commented:
Pushing two commits to measure the sidebar and full modal widths both in em, which I confirmed looks okay at 12px to 20px.

sahil839 reviewed (commented):
Posted one comment.

On 12px, I think tabs can be aligned centrally with the left panel -
![Screenshot from 2025-01-10 12-51-14](https://github.com/user-attachments/assets/41f47d12-0807-4000-9244-e30ee3af8e73)

On 20px, I was thinking we should increase the left panel width so that tabs can be aligned centrally but as you mentioned the space for right panel will be decreased so not sure about that.
![Screenshot from 2025-01-10 12-51-52](https://github.com/user-attachments/assets/2b7035d6-057c-4af8-a0bc-d3177ef0b2dc)

alya commented:
Yeah, it doesn't necessarily have to be on this PR, but widening the left side at 20px is good. We could at least avoid this in English:
![Screenshot 2025-01-09 at 23 47 30@2x](https://github.com/user-attachments/assets/7d873187-1eaf-4ebf-8417-8bd8e878f804)

It would probably make sense to also increase the max width of the overlay as a whole, and to collapse the left panel sooner, to avoid squishing the right side too much.

alya commented:
Ooh, yeah, Russian gets quite bad at 20px with the current width:
![Screenshot 2025-01-09 at 23 49 40@2x](https://github.com/user-attachments/assets/555ac886-20e2-4e0e-8934-dee5969d3146)

evykassirer commented:
> It would probably make sense to also increase the max width of the overlay as a whole

That's in this PR! Try it out and let me know what you think :)

evykassirer commented:
> On 20px, I was thinking we should increase the left panel width so that tabs can be aligned centrally but as you mentioned the space for right panel will be decreased so not sure about that.

It seems like if I remove the max-size, it still looks fine -- due to the increasing modal size and the early collapsing change below. So now the sidebar is always as wide as the tabs (or rather, slightly wider).

> I think tabs can be aligned centrally with the left panel

Ah yes, I removed some absolute positioning to let it center naturally with flex, thanks!

> collapse the left panel sooner, to avoid squishing the right side too much.

Great idea -- I changed it to collapse at 1000px at 14px, scaling with font size. Note this also changes the subscriptions modal, since they share this styling.

## some updated screenshots 

20px:

<img width="1344" alt="image" src="https://github.com/user-attachments/assets/ba804123-a042-43dd-a876-79e149f3fadc" />

<img width="1340" alt="image" src="https://github.com/user-attachments/assets/39533f00-f892-4d28-9de4-39db722d1fdd" />


12px:

<img width="1338" alt="image" src="https://github.com/user-attachments/assets/f58737d4-a255-41d6-9c9a-c652484ebf45" />

<img width="1337" alt="image" src="https://github.com/user-attachments/assets/78fa5940-5e6d-4946-baa4-93c65dec8d51" />

sahil839 commented on web/src/css_variables.ts:
```diff
@@ -37,6 +37,7 @@ export const media_breakpoints = {
     cb4_min: cb4 + "px",
     cb5_min: cb5 + "px",
     short_navbar_cutoff_height: "600px",
+    modal_sidebar_collapse_breakpoint: 1000 / 14 + "em",
```
Maybe `settings_overlay_sidebar_collapse_breakpoint` would be more clear here.

evykassirer replied:
Sure!

evykassirer replied:
Actually, the reason I did it more broadly phrased is that it's also used for the subscriptions modal. Those modals share a lot of CSS.

sahil839 replied:
We can name it as `overlay_sidebar_collapse_breakpoint` also rather than using modal. Or we also refer to subscriptions UI as stream settings or channel settings as well, so having `settings` should also be fine.

sahil839 reviewed (commented):
Looks good overall. Not sure of the new breakpoint for collapsing the left panel, I guess we can delay the collapsing some more but might to make sure that it looks fine for different languages.

evykassirer commented:
> make sure that it looks fine for different languages

It seems like Russian was particularly long so I shared some screenshots above with that. Are there other languages in particular worth testing?

evykassirer commented:
@alya I think it's worth you taking another look at this point

alya commented:
Looks like there's a regression where the header area gets shorter in narrow windows. E.g., narrow window at 20px:

![Screenshot 2025-01-22 at 13 28 16@2x](https://github.com/user-attachments/assets/3a4bc95d-d1e7-44ba-940b-6f3cbc83a8f5)

alya commented:
It feels like the left-side panel collapses way too early at small font sizes.

![Screenshot 2025-01-22 at 13 30 28@2x](https://github.com/user-attachments/assets/d2f2c810-b1e5-4a4c-8bc4-6cb392ad06f7)

alya commented:
Seems OK to me otherwise.

sahil839 commented:
> It seems like Russian was particularly long so I shared some screenshots above with that. Are there other languages in particular worth testing?

I did not had any particular language in mind but thought we would need to test the breakpoint with different languages. I tested and it seems like the texts are not longer in languages other than Russian.

sahil839 commented:
I do not have any concerns other than the ones Alya already mentioned.

karlstolley commented on web/styles/settings.css:
```diff
@@ -1308,20 +1308,22 @@ label.preferences-radio-choice-label {
                     position: absolute;
                     display: flex;
-                    left: 10px;
+                    left: 0.7142em; /* 10px at 14px em */
```
Is the layout going to tolerate being shifted left by an em value? My feeling is that we should tread carefully on horizontal matters (padding, width, left/right offsets, etc.)

evykassirer replied:
This left margin and absolute positioning have been removed -- it's now in flexbox and centers itself horizontally in the left width space

evykassirer replied:
I'll just avoid changing it in this commit, since it gets removed later.

karlstolley commented on web/styles/settings.css:
```diff
@@ -1,7 +1,7 @@
 $settings_header_height: 3.2142em; /* 45px / 14px em */
 /* The width of the settings sidebar. */
-$settings_sidebar_width: 255px;
+$settings_sidebar_width: 18.2142em; /* 255px / 14px em */
```
Is this the right away to go? Or would it be better to express as a percentage?

karlstolley replied:
Okay. Also, could you note your px-to-em conversions like so: `255px at 14px/1em`?

evykassirer replied:
Oh yes, sorry I probably opened this PR before I had a convention for this.

karlstolley commented on web/styles/settings.css:
```diff
@@ -1272,7 +1272,7 @@ label.preferences-radio-choice-label {
     height: 95vh;
     width: 97vw;
-    max-width: 1024px;
+    max-width: 73.1428em;
```
This conversion is probably okay, but I'd like to have the pixel value for the em notated here.

evykassirer replied:
Oops, thanks for noticing

karlstolley commented on web/styles/settings.css:
```diff
@@ -1316,9 +1316,7 @@ label.preferences-radio-choice-label {
-                    position: absolute;
                     display: flex;
-                    left: 0.7142em; /* 10px at 14px em */
                     z-index: 1;
```
With positioning gone, do we still have need for the `z-index` value here?

evykassirer replied:
Seems like it can be removed, yeah, thanks!

karlstolley commented on web/src/css_variables.ts:
```diff
@@ -37,6 +37,7 @@ export const media_breakpoints = {
     cb4_min: cb4 + "px",
     cb5_min: cb5 + "px",
     short_navbar_cutoff_height: "600px",
+    modal_sidebar_collapse_breakpoint: 1000 / 14 + "em",
```
Can we leave a comment describing where the `1000 / 14` values are originating from?

evykassirer replied:
Added "// 1000px at legacy 14px font-size, scaled with em to user-chosen font-size"

evykassirer commented:
> Looks like there's a regression where the header area gets shorter in narrow windows. E.g., narrow window at 20px:

I've fixed this and it also brought me on a bit of a rabbithole where I updated several other things for all three settings menu like this (user groups, subscriptions, main settings), so it would be good to manually test all three of those to double check headers look okay.

> It feels like the left-side panel collapses way too early at small font sizes.

Oh yeah the breakpoint isn't really updating properly. I'm going to ask someone (Anders or Karl probably) for help looking into this. Thanks for pointing that out.

evykassirer commented:
I believe all comments are addressed now except for the breakpoint issues.

evykassirer commented:
Fixed the breakpoint issues with @container.

karlstolley commented on web/styles/subscriptions.css:
```diff
@@ -784,6 +775,12 @@ h4.user_group_setting_subsection_title {
+#settings_page,
+#groups_overlay,
+#subscription_overlay {
+    container: settings_overlay / inline-size;
```
Two things here: 1, let's style container names with hyphens (`settings-overlay`), and 2, let's be sure to mention the container name in `@container` queries. That'll save us a ton of pain in the future, so we don't have to hunt through the DOM and CSS to figure out where a container is specified. So, `@container settings-overlay (width > $something)`.

evykassirer replied:
Done and done, thanks!

karlstolley commented on web/styles/settings.css:
```diff
@@ -1,8 +1,6 @@
-    --settings-header-height: 45px;
     /* The width of the settings sidebar. */
-    --settings-sidebar-width: 255px;
+    --settings-sidebar-width: 18.2142em; /* 255px at 14px/em */
```
Is there a reason this variable can't live with the others in `app_variables.css`?

evykassirer replied:
Mostly I didn't move it because I didn't have to and was minimizing changes. I can add another commit moving it.

karlstolley commented on web/src/css_variables.ts:
```diff
@@ -37,6 +37,11 @@ export const media_breakpoints = {
+    // measured as the width of the overlay itself, not the width of the full
+    // screen. 800px is the breakpoint at the 14px legacy font size, scaled with
+    // em to user-chosen font-size.
+    settings_overlay_sidebar_collapse_breakpoint: 800 / 14 + "em",
```
I have a commit in #33249 that introduces a new area of this file for container-query breakpoints. If that merges, it might be nice to move this there.

karlstolley commented on web/styles/settings.css:
```diff
@@ -1286,7 +1284,7 @@ label.preferences-radio-choice-label {
     height: 95vh;
     width: 97vw;
-    max-width: 1024px;
+    max-width: 73.1428em; /* 1024px at 14px em */
```
I'd round this up to `73.1429em` (since it works out to 73.14285...)

evykassirer replied:
Sure

karlstolley commented on web/styles/settings.css:
```diff
@@ -1321,40 +1319,37 @@ label.preferences-radio-choice-label {
-            @media (width < $md_min) {
+            @container (width >= $settings_overlay_sidebar_collapse_breakpoint) {
+                .tab-switcher .ind-tab {
+                    min-width: 6.7857em; /* 95px at 14px em */
```
Would there be any harm in moving this `min-width` value out of a container query? possibly with the `justify-content` below? How is the content justified outside of the smaller width?

evykassirer replied:
I tried both of these rules at both widths and it seems fine/better to have them for both! Thanks for noticing.

karlstolley commented on web/styles/settings.css:
```diff
@@ -1790,7 +1778,7 @@ label.preferences-radio-choice-label {
 
 /* This value needs to match with the same in subscriptions.css, as
```
Should this comment be removed or updated, given the movement of styles out of the `subscriptions.css` file?

evykassirer replied:
I don't think so, since this is still present in `subscriptions.css`

```
/* Note that this block has settings_page CSS as well, and thus needs
   to match the media queries in settings.css.

   Longer-term we should extract this logic two-column-overlay class
   to read more naturally. */
```

karlstolley commented on web/styles/settings.css:
```diff
@@ -1835,7 +1840,10 @@ label.preferences-radio-choice-label {
             width: 100%;
             border: none;
             /* 48px is the height of settings header and 45px is the height of tab-container */
```
I think this comment can be struck, too, since the variables do the talking about what's happening now :)

evykassirer replied:
Totally! Thanks

karlstolley commented on web/styles/settings.css:
```diff
@@ -1846,7 +1854,7 @@ label.preferences-radio-choice-label {
-@media (width < 830px) and (width > $md_min) {
+@container (width < 830px) and (width > $settings_overlay_sidebar_collapse_breakpoint) {
     #attachments-settings .upload-file-name {
         min-width: 10em;
```
Just studying the code again, but I wonder why we need such a complex query for setting a `min-width`? Having one condition here in px and the other in ems makes this really hard to think about.

evykassirer replied:
Looking at attachment settings as an example, it seems like the max cut-off isn't really necessary, we could just do a min width for this. It's just guaranteeing that columns get enough space, and with wider screens then there's even more space.

evykassirer replied:
Added a commit for that at the start of this PR

evykassirer commented:
I believe every comment has been addressed or responded to. @karlstolley back to you :)

karlstolley commented:
Cool. I haven't re-tested, but I think this looks good to merge, especially since it'll unblock some of @evykassirer's other work. It wouldn't be shocking to have some follow-ups on the settings, but this is good to go, I believe.

timabbott commented:
Merged, thanks @evykassirer and @karlstolley!
